src/clients/github.js

```diff
@@ -0,0 +1,280 @@
+  if (project.hiddenUIComponents.length) {
+    json.hiddenUIComponents = project.hiddenUIComponents;
+  }
+  return JSON.stringify(json);
+}
+
+function _performWithRetryNetworkErrors(perform, options = {}) {
+  return performWithRetries(
+    perform,
+    errorMessage => errorMessage === NETWORK_ERROR,
+    options,
+  );
+}
+
+function _performWithRetryNotFoundOrNetworkErrors(perform) {
+  const retryErrors = [NETWORK_ERROR, 'NOT_FOUND'];
+  return performWithRetries(
+    perform,
+    errorMessage => retryErrors.includes(errorMessage),
+  );
+}
+
+async function _createSourceFile(
+  github, userName, repoName, branchName, fileName, source) {
+  return github.getRepo(userName, repoName).
+    writeFile(
+      branchName,
+      fileName,
+      source,
+      COMMIT_MESSAGE,
+      {},
+    );
+}
+
+async function _createHtmlFile(github, project, userName, repoName) {
+  if (trim(project.sources.html)) {
+    return _createSourceFile(
+      github,
+      userName,
+      repoName,
+      MASTER,
+      'raw_index.html',
+      project.sources.html,
+    );
+  }
+  return Promise.resolve();
+}
+
+async function _createCssFile(github, project, userName, repoName) {
+  if (trim(project.sources.css)) {
+    return _createSourceFile(
+      github,
+      userName,
+      repoName,
+      MASTER,
+      'styles.css',
+      project.sources.css,
+    );
+  }
+  return Promise.resolve();
+}
+
+async function _createJsFile(github, project, userName, repoName) {
+  if (trim(project.sources.javascript)) {
+    return _createSourceFile(
+      github,
+      userName,
+      repoName,
+      MASTER,
+      'script.js',
+      project.sources.js,
+    );
+  }
+  return Promise.resolve();
+}
+
+async function _createPreviewFile(github, project, userName, repoName) {
+  const preview = generatePreview(project);
+  return _createSourceFile(
+    github,
+    userName,
+    repoName,
+    GH_PAGES,
+    'index.html',
+    preview,
+  );
+}
+
+async function _createBranch(
+  github,
+  userName,
+  repoName,
+  oldBranch,
+  newBranch) {
+  return github.getRepo(userName, repoName).
+    createBranch(oldBranch, newBranch);
+}
+
+async function _updateRepoDescription(github, userName, repoName) {
+  const url = `https://${userName}.github.io/${repoName}`;
+  return github.getRepo(userName, repoName).
+    updateRepository({
+      name: repoName,
+      description: url,
+      homepage: url,
+    });
+}
+
+function _githubWithAccessToken(token) {
+  return new GitHub({auth: 'oauth', token});
+}
+
+function _getGithubToken(user) {
+  return get(user, ['accessTokens', 'github.com']);
+}
+
+function _clientForUser(user) {
+  const githubToken = _getGithubToken(user);
+  if (githubToken) {
+    return _githubWithAccessToken(githubToken);
+  }
+
+  return _anonymousGithub;
+}
+
+function _canUpdateGist(user) {
+  return Boolean(_getGithubToken(user));
+}
+
+export {
+  createGistFromProject,
+  createRepoFromProject,
+  EmptyGistError,
+  loadFromId,
+};
```

Comment:
I know this style is still used in some places in the code, but my preference is for the `export` statements to go inline, e.g.:

```js
export async function createGistFromProject() {
  …
```

(That solves the problem of being able to easily tell which functions are part of the module’s public interface without needing the `_` notation)